### PR TITLE
SD: Types, traits, and expression class

### PR DESCRIPTION
**Pull request by jppelteret**

### CODE REVIEW

File: tests/symengine/symengine_number_type_01.output [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+DEAL:Double:Math operators with normal numbers::c+=s: 7.0
+DEAL:Double:Math operators with normal numbers::c*=s: 14.0
+DEAL:Double:Math operators with normal numbers::c-=s: 12.0
+DEAL:Double:Math operators with normal numbers::c/=s: 6.0
+DEAL:Double:Math operators with normal numbers::x: x
+DEAL:Double:Math operators with symbols::c+=x: 6.0 + x
+DEAL:Double:Math operators with symbols::c*=x: x*(6.0 + x)
+DEAL:Double:Math operators with symbols::c-=x: -x + x*(6.0 + x)
+DEAL:Double:Math operators with symbols::c/=x: (-x + x*(6.0 + x))/x
+DEAL:Double:Differentiation::dc_dx: x*(-1 + x + 6.0 + x) - x + x*(6.0 + x)
```
Comment (isuruf) [inline tests/symengine/symengine_number_type_01.output]: the derivative here looks wrong

Comment (jppelteret) [reply tests/symengine/symengine_number_type_01.output]: Wow, lucky catch! I still needed to update this tests's blessed output, which would have overwritten this :-)

**Comment by jppelteret** [comment]
@luca-heltai You might find bbee6ab interesting... With this new constructor we will be able to support piecewise defined symbolic functions (so like the teniary operator and if-else if-else statements). You can check out the [test here](https://github.com/dealii/dealii/commit/bbee6ab2e6b280c2b5d4b476f0226833d2ef92e3#diff-1acb3fdaeb6141717f6d9272169baf4bR737) to see an example of this.

**Comment by jppelteret** [comment]
@luca-heltai If you've got some time, this is ready for some initial feedback. Some necessary patches have been accepted into SymEngine, so this could now compile and run against their `master` branch.

File: include/deal.II/differentiation/sd/symengine_number_traits.h [hunk 2]
```diff
@@ -0,0 +1,95 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2016 - 2019 by the deal.II authors
```
Comment (luca-heltai) [inline include/deal.II/differentiation/sd/symengine_number_traits.h]: same here

File: include/deal.II/differentiation/sd/symengine_number_traits.h [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+{
+  namespace SD
+  {
+    /**
+     * A struct to indicate whether a given @p NumberType is a supported
+     * symbolically differentiable number or not. By default, numbers are not
+     * considered to have the necessary characteristics to fulfill this
+     * condition.
+     *
+     * @author Jean-Paul Pelteret, 2017
```
Comment (luca-heltai) [inline include/deal.II/differentiation/sd/symengine_number_traits.h]: Update the dates to 2019

File: include/deal.II/differentiation/sd/symengine_number_traits.h [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+    {};
+
+
+    /**
+     * A struct to indicate whether a given @p NumberType is a supported
+     * SymEngine number or not. By default, numbers are not
+     * considered to have the necessary characteristics to fulfill this
+     * condition.
+     *
+     * @author Jean-Paul Pelteret, 2017
```
Comment (luca-heltai) [inline include/deal.II/differentiation/sd/symengine_number_traits.h]: same here

File: include/deal.II/differentiation/sd/symengine_number_traits.h [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+     * @author Jean-Paul Pelteret, 2019
+     */
+    template <>
+    struct is_symengine_number<SymEngine::Expression> : std::true_type
+    {};
+
+
+    /**
+     * A struct to indicate whether a given @p NumberType is a supported
+     * SymEngine number or not.
```
Comment (luca-heltai) [inline include/deal.II/differentiation/sd/symengine_number_traits.h]: symbolically differentiable number or not.

File: include/deal.II/differentiation/sd/symengine_number_types.h [hunk 6]
```diff
@@ -0,0 +1,1036 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2016 - 2019 by the deal.II authors
```
Comment (luca-heltai) [inline include/deal.II/differentiation/sd/symengine_number_types.h]: 2019

File: include/deal.II/differentiation/sd/symengine_number_types.h [hunk 7]
```diff
@@ -0,0 +1,1036 @@
...
+    /**
+     * A class to wrap SymEngine expressions.
+     *
+     * With this number type, SymEngine numbers can be used to perform scalar
+     * and tensor mathematics in deal.II. It (or the SymEngine::Expression
+     * class, of which it stores an instance) therefore forms the basis of
+     * symbolic computation via SymEngine in deal.II. With it one can
+     * perform symbolic differentiation and subsequent substitution with both
+     * scalars and deal.II's native Tensor and SymmetricTensor types.
+     *
```
Comment (luca-heltai) [inline include/deal.II/differentiation/sd/symengine_number_types.h]: Maybe add a small example here, on how to use this class?

**Comment by jppelteret** [comment]
After a little bit of effort, I've managed to incorporate load/save/serialisation capabilities into the class as well. This will be really useful, as it means that one can pre-compute residuals, linearisation etc and save them for later use. I'll try to maintain this functionality in as many places as would be possible.

File: include/deal.II/differentiation/sd/symengine_number_types.h (hunk 7 above)
Comment (jppelteret) [reply include/deal.II/differentiation/sd/symengine_number_types.h]: I've added a number of examples, which I'm sure should suffice :-)

**Comment by jppelteret** [comment]
I guess I've gone as far as I want to go with this PR. If I find any missing functionality I can always add it later. Next up would be the definition of more math functions in this namespace.

If there are no further comments on this in the next few days, I'll squash all of the commits into a couple of logical commits.

**Comment by jppelteret** [comment]
From my point of view, I think that this is good to go.

**Pull request merged**